core/handler/manager_server.go

```diff
@@ -73,6 +73,60 @@ func (h *handlerManager) validateTTNAuthAppContext(ctx context.Context, appID st
 	return ctx, claims, nil
 }
 
+func (h *handlerManager) SetRegisterOnJoin(ctx context.Context, in *pb_lorawan.SetRegisterOnJoinMessage) (*empty.Empty, error) {
+	if err := in.Validate(); err != nil && !errors.IsNotFound(err) {
+		return nil, errors.Wrap(err, "Invalid SetRegisterOnJoin message request")
+	}
+	ctx, claims, err := h.validateTTNAuthAppContext(ctx, in.AppId)
+	if err != nil {
+		return nil, err
+	}
+	err = checkAppRights(claims, in.AppId, rights.Devices)
+	if err != nil {
+		return nil, err
+	}
+
+	var app *application.Application
+	if app, err = h.handler.applications.Get(in.AppId); err != nil {
+		return nil, errors.Wrap(err, "Application not registered to this Handler")
+	}
+
+	app.StartUpdate()
+
+	app.OnJoinRegistration = in.Val
+	if in.Val {
+		app.OnJoinRegistrationAppEui = *in.AppEui
+		app.OnJoinRegistrationAppKey = *in.AppKey
+		app.OnJoinRegistrationAccessKey = in.AccessKey
+		app.OnJoinRegistrationAccessKeyName = in.AccessKeyName
+	} else {
+		in.AppEui = &app.OnJoinRegistrationAppEui
+	}
+	h.handler.Ctx.WithFields(ttnlog.Fields{
+		"AppEui":        app.OnJoinRegistrationAppEui,
+		"AppKey":        app.OnJoinRegistrationAppKey,
+		"AccessKey":     app.OnJoinRegistrationAccessKey,
+		"AccessKeyName": app.OnJoinRegistrationAccessKeyName,
+	}).Debug("Registering on-join parameters...")
+
+	err = h.handler.applications.Set(app)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = h.deviceManager.SetRegisterOnJoin(ctx, in)
+	if err != nil {
+		// Try to revert the changes
+		app.OnJoinRegistration = !in.Val
+		if err := h.handler.applications.Set(app); err != nil {
+			h.handler.Ctx.WithField("AppID", app.AppID).Error("Failed to revert the changes at the handler level when SetRegisterOnJoin failed")
```

Comment:
Also it might be better to set it locally only when it has been set successfully in the Broker